File: app/android/android/app/src/main/java/com/gagagps/operator/rtk/NtripSourceTable.kt
package com.gagagps.operator.rtk

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.net.Socket
import java.net.SocketTimeoutException

// una entrada "STR;..." de la sourcetable NTRIP - formato estandar (RTCM Ntrip 1/2), el mismo
// que cualquier caster real expone al pedir la raiz ("/") en vez de un mountpoint especifico
data class NtripMountpoint(
    val mountpoint: String,
    val identifier: String,
    val format: String,
    val navSystem: String,
    val country: String,
    val latitude: Double?,
    val longitude: Double?,
    val nmeaRequired: Boolean,
)

// GET / (sin mountpoint) contra cualquier caster NTRIP real regresa la lista completa de puntos
// de montura disponibles - estandar del protocolo, no algo especifico de un proveedor. Formato:
// lineas "STR;mountpoint;identifier;format;format-details;carrier;nav-system;network;country;
// lat;lon;nmea;solution;generator;compr-encr;auth;fee;bitrate;..." hasta "ENDSOURCETABLE".
object NtripSourceTable {
    suspend fun fetch(host: String, port: Int): List<NtripMountpoint> = withContext(Dispatchers.IO) {
        val socket = Socket(host, port)
        socket.soTimeout = 10000
        try {
            val request = "GET / HTTP/1.1\r\n" +
                "Host: $host\r\n" +
                "User-Agent: NTRIP GagaOperador/1.0\r\n" +
                "Connection: close\r\n\r\n"
            socket.getOutputStream().write(request.toByteArray())
            socket.getOutputStream().flush()

            val input = socket.getInputStream()
            val statusLine = readNtripLine(input)
            if (statusLine == null || (!statusLine.contains("200") && !statusLine.contains("SOURCETABLE"))) {
                throw Exception("El caster no respondio con la tabla de fuentes: ${statusLine ?: "sin respuesta"}")
            }
            while (true) {
                val header = readNtripLine(input) ?: break
                if (header.isEmpty()) break
            }

            val mountpoints = mutableListOf<NtripMountpoint>()
            while (true) {
                val row = readNtripLine(input) ?: break
                if (row.startsWith("ENDSOURCETABLE")) break
                if (!row.startsWith("STR;")) continue
                val f = row.split(";")
                if (f.size < 9) continue
                mountpoints.add(
                    NtripMountpoint(
                        mountpoint = f.getOrElse(1) { "" },
                        identifier = f.getOrElse(2) { "" },
                        format = f.getOrElse(3) { "" },
                        navSystem = f.getOrElse(6) { "" },
                        country = f.getOrElse(8) { "" },
                        latitude = f.getOrNull(9)?.toDoubleOrNull(),
                        longitude = f.getOrNull(10)?.toDoubleOrNull(),
                        nmeaRequired = f.getOrNull(11) == "1",
                    ),
                )
            }
            mountpoints
        } catch (e: SocketTimeoutException) {
            throw Exception("Tiempo de espera agotado consultando el caster")
        } finally {
            try {
                socket.close()
            } catch (_: Exception) {
            }
        }
    }
}

File: backend/src/utils/ntripSourceTable.ts
import net from 'net';

// misma logica que NtripSourceTable.kt (app/android/.../rtk/NtripSourceTable.kt) - duplicada a
// proposito, no hay forma de compartir codigo real entre Node y Kotlin (ver gotcha de
// shared-types en CLAUDE.md). GET / (sin mountpoint) contra cualquier caster NTRIP real regresa
// la tabla completa de fuentes disponibles - estandar del protocolo (RTCM Ntrip 1/2), no algo
// especifico de un proveedor. Formato de cada fila: "STR;mountpoint;identifier;format;
// format-details;carrier;nav-system;network;country;lat;lon;nmea;..." hasta "ENDSOURCETABLE".
export interface NtripMountpointInfo {
  mountpoint: string;
  identifier: string;
  format: string;
  navSystem: string;
  country: string;
  latitude: number | null;
  longitude: number | null;
  nmeaRequired: boolean;
}

const SOCKET_TIMEOUT_MS = 10000;

export function fetchNtripSourceTable(host: string, port: number): Promise<NtripMountpointInfo[]> {
  return new Promise((resolve, reject) => {
    let buffer = '';
    let state: 'status' | 'headers' | 'body' = 'status';
    let settled = false;
    const mountpoints: NtripMountpointInfo[] = [];

    const socket = net.connect({ host, port });

    const finish = () => {
      if (settled) return;
      settled = true;
      socket.destroy();
      resolve(mountpoints);
    };
    const fail = (err: Error) => {
      if (settled) return;
      settled = true;
      socket.destroy();
      reject(err);
    };

    socket.setTimeout(SOCKET_TIMEOUT_MS, () => fail(new Error('Tiempo de espera agotado consultando el caster')));
    socket.on('error', (err) => fail(err));
    socket.on('close', finish);

    socket.on('connect', () => {
      socket.write(
        `GET / HTTP/1.1\r\nHost: ${host}\r\nUser-Agent: NTRIP GAGA-GPS/1.0\r\nConnection: close\r\n\r\n`,
      );
    });

    socket.on('data', (chunk) => {
      // latin1: la tabla de fuentes es ASCII/latin1 puro, nunca UTF-8 multibyte - igual que el
      // parser NMEA del lado Kotlin
      buffer += chunk.toString('latin1');
      let idx: number;
      while ((idx = buffer.indexOf('\n')) >= 0) {
        const line = buffer.slice(0, idx).replace(/\r$/, '');
        buffer = buffer.slice(idx + 1);

        if (state === 'status') {
          if (!line.includes('200') && !line.includes('SOURCETABLE')) {
            fail(new Error(`El caster no respondió con la tabla de fuentes: ${line || 'sin respuesta'}`));
            return;
          }
          state = 'headers';
          continue;
        }
        if (state === 'headers') {
          if (line.trim() === '') state = 'body';
          continue;
        }
        if (line.startsWith('ENDSOURCETABLE')) {
          finish();
          return;
        }
        if (!line.startsWith('STR;')) continue;
        const f = line.split(';');
        if (f.length < 9) continue;
        mountpoints.push({
          mountpoint: f[1] ?? '',
          identifier: f[2] ?? '',
          format: f[3] ?? '',
          navSystem: f[6] ?? '',
          country: f[8] ?? '',
          latitude: f[9] ? Number(f[9]) : null,
          longitude: f[10] ? Number(f[10]) : null,
          nmeaRequired: f[11] === '1',
        });
      }
    });
  });
}
